src/solver.rs

```diff
@@ -93,50 +92,57 @@ pub fn resolve<P: Package, V: Version>(
             .should_cancel()
             .map_err(|err| PubGrubError::ErrorShouldCancel(err))?;
 
-        state.unit_propagation(next)?;
+        state.unit_propagation(next.clone())?;
 
-        // Pick the next package.
-        let (p, term) = match state.partial_solution.pick_package(dependency_provider)? {
-            None => {
-                return state.partial_solution.extract_solution().ok_or_else(|| {
-                    PubGrubError::Failure(
-                        "How did we end up with no package to choose but no solution?".into(),
-                    )
-                })
-            }
-            Some(x) => x,
-        };
-        next = p.clone();
-        let available_versions =
-            dependency_provider
-                .list_available_versions(&p)
-                .map_err(|err| PubGrubError::ErrorRetrievingVersions {
-                    package: p.clone(),
-                    source: err,
-                })?;
+        let potential_packages = state.partial_solution.potential_packages();
+        if potential_packages.is_none() {
+            drop(potential_packages);
+            // The borrow checker did not like using a match on iter.
+            // This `if ... is_none ... drop` is a workaround.
+            // I believe this is a case where Polonius could help, when and if it lands in rustc.
+            return state.partial_solution.extract_solution().ok_or_else(|| {
+                PubGrubError::Failure(
+                    "How did we end up with no package to choose but no solution?".into(),
+                )
+            });
+        }
+        let decision = dependency_provider
+            .make_decision(potential_packages.unwrap())
+            .map_err(PubGrubError::ErrorMakeDecision)?;
+        next = decision.0.clone();
 
         // Pick the next compatible version.
-        let v = match PartialSolution::<P, V>::pick_version(&available_versions[..], &term) {
+        let v = match decision.1 {
             None => {
+                let term = state.partial_solution.term_intersection_for_package(&next);
                 state.add_incompatibility(|id| {
-                    Incompatibility::no_version(id, p.clone(), term.clone())
+                    Incompatibility::no_version(id, next.clone(), term.clone())
                 });
                 continue;
             }
             Some(x) => x,
         };
+        if !state
+            .partial_solution
+            .term_intersection_for_package(&next)
+            .contains(&v)
+        {
+            return Err(PubGrubError::Failure(
```

Comment:
Should we return a newly introduced error here?

Reply:
or just panic, it is not clear to me how a user of our code could use a `match` in this case. Like the only fix is to fix there implementation.

Reply:
I'm not sure what panic brings. With an error, the caller can do whatever they want when a `Failure` happens (they could panic). And new error wouldn't bring much either since they have to fix their bug anyway.

Reply:
AFAIK panicking should be done on internal library bugs, which the user can do nothing about. There are a few we define as Failures currently, which I wanted to change to panics in some future PR.

As for user introduced bugs... I don't know actually what the consensus is. I just changed the error to be more specific in the latest commit for now, but I'm open to the discussion on the topic.